### PR TITLE
Watch UI v2

### PR DESCRIPTION
**Pull request by janegermaier**
# WatchUI ver2.0.1
## Breaking changes
WatchUI 2.0 brings breaking changes. Dev team decided to streamline the library and focus it solely on the image and text comparison. This allows us to remove the implicit dependency on browser automation libraries - namely SeleniumLibrary, which was implicit part of the library via RF BuiltIn() import of the SeleniumLibrary instance.

This is no longer the case - user of the WatchUI therefore can and have to choose, what UI automation library will use and provide screenshots to the WatchUI keywords to be compared. It could be now user with SeleniumLibrary, Browser library, Sikuli, Appium or any other UI library where visual validation is required.

Version 1.x.x is no longer supported, but it is still available on [Pypi](pip install WatchUI==1.0.11).

## ChangeLog

- Delete Selenium from codes (more in Breaking changes)
  - Delete keyword Compare Screen
  - Delete keyword Create Screens
  - Delete keyword Compare Screen Get Information
- Change in keyword Compare Screen Without Areas. Keyword doesnt use selenium and if u want to use it, you must have two exists screen in folder and compare them.
- Better loggin messages (We use robot api)
- New documentation
- Changes in readme.md
- Changes in tests
- Add example folder with examples of How to use library
- In test folder is just unit test with Robot Framework
- Add unit test for methods (Thanks @radekBednarik )
- We have no longer one large file with all keywords but now keywords are split in the small files (folder keywords) and imported into WatchUI.py
- Repair import library settings in robot files.

### CODE REVIEW

File: .github/workflows/code-formatter.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        id: git-check
+        run: echo ::set-output name=modified::$(if git diff-index --quiet HEAD --; then echo "false"; else echo "true"; fi)
+      - name: Push changes
+        if: steps.git-check.outputs.modified == 'true'
+        run: |
+          git config --global user.name 'radekBednarik'
+          git config --global user.email 'radek.bednarik@tesena.com'
+          git remote set-url origin https://x-access-token:${{ secrets.GITHUB_TOKEN }}@github.com/${{ github.repository }}
+          git commit -am "code autoformat via Black"
+          git push
```
Comment (radekBednarik) [inline .github/workflows/code-formatter.yml:26]: add newline

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -1,53 +1,53 @@
...
+  pull_request:
+    branches: [master]
+
+# A workflow run is made up of one or more jobs that can run sequentially or in parallel
+jobs:
+  robot_test:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: [3.6]
```
Comment (radekBednarik) [inline .github/workflows/main.yml:15]: update the Python version to 3,9+, some of the source code is typed, so 3.6 does not make sense

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -1,53 +1,53 @@
...
+            && unzip chromedriver_linux64.zip \
+            && rm -rf chromedriver_linux64.zip \
+            && sudo mv chromedriver /usr/local/bin/chromedriver \
+            && sudo chmod +x /usr/local/bin/chromedriver \
+            && chromedriver --version
+      - name: display
+        run: DISPLAY=:0
+      - name: Robot Framework
+        run: |
+          python -m pip install --upgrade pip
```
Comment (radekBednarik) [inline .github/workflows/main.yml:38]: or:
pip install pipx
pipx install poetry
poetry install

that way you do not have to check that you have all the packages here, when they are in the pyproject.toml

File: .github/workflows/python-publish.yml [hunk 4]
```diff
@@ -1,30 +1,29 @@
...
+        python -m pip install --upgrade pip
+    - name: Install poetry
+      run: |
+        pip install poetry
+    - name: Build package
+      run: |
+        poetry build
+    - name: Publish package
+      run: |
+        poetry publish -u ${{secrets.PYPI_USERNAME}} -p ${{secrets.PYPI_PASSWORD}}
```
Comment (radekBednarik) [inline .github/workflows/python-publish.yml:29]: add newline

File: LICENSE [hunk 5]
```diff
@@ -1,21 +1,21 @@
...
-
-THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+MIT License
+
+Copyright (c) 2020 Jan Egermaier, Radek Bednarik
```
Comment (radekBednarik) [inline LICENSE:3]: add other team members :)

File: README.md [hunk 6]
```diff
@@ -1,42 +1,54 @@
...
+
+[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
+
+## [Documentation](https://tesena-smart-testing.github.io/WatchUI/) | [Tesena](https://www.tesena.com/) | [Pypi](https://pypi.org/project/WatchUI/)
+
+## Important notice for users
+
+WatchUI 2.0 brings breaking changes. Dev team decided to streamline the library and focus it solely on the image and text comparison. This allows us to remove the implicit dependency on browser automation libraries - namely SeleniumLibrary, which was implicit part of the library via RF `BuiltIn()` import of the SeleniumLibrary instance.
+
+This is no longer the case - user of the WatchUI therefore **can and have to choose, what UI automation library will use** and provide screenshots to the WatchUI keywords to be compared. It could be now user with SeleniumLibrary, Browser library, Sikuli, Appium or any other UI library where visual validation is required.
```
Comment (radekBednarik) [inline README.md:11]: typo in "user" => "used"

File: WatchUI/Ibasic/IBasic.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+            type_of_messages (Literal[, optional): Type of logging message. Defaults to "Info".
+            path_to_image (str, optional): Path to image to be appended to log message.
+            Defaults to "".
+        Raises:
+            AssertionError:
+        """
+        if work_object == "Image":
+            if type_of_messages == "Info":
+                logger.info(
+                    "Image was saved to your output folder </p><img src='"
```
Comment (radekBednarik) [inline WatchUI/Ibasic/IBasic.py:118]: why not string templates?

File: WatchUI/Ibasic/IBasic.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+                    + path_to_image
+                    + "'/>",
```
Comment (radekBednarik) [inline WatchUI/Ibasic/IBasic.py:120]: why not string templates?

File: WatchUI/WatchUI.py [hunk 9]
```diff
@@ -1,807 +1,63 @@
...
+                format_image {str} -- Format for saving picture/screenshot (png, jpg etc.)
+                tesseract_path {str} -- path for tesseract
+                Example: format_image=jpg (default: png)
+            Examples:
+            | =Setting= | =Value= | =Value= | =Value= | =Comment= |
+            | Library   | WatchUI |      |  | # Uses default values of keyword arguments |
+            | Library   | WatchUI | outputs_folder=<path_to_folder> | | # changes folder to different one |
+            | Library   | WatchUI | outputs_folder=<path_to_folder> | ssim_basic=<float> | # changes output folder and ssim threshold |
+            """
+        super(Image, self).__init__(
```
Comment (radekBednarik) [inline WatchUI/WatchUI.py:58]: super().__init() should be called inside the subclass __init__() method, or am I mistaken? no wonder, that pylint threw the error, which you supressed ;)

File: WatchUI/__init__.py [hunk 10]
```diff
@@ -1 +1 @@
-from WatchUI.WatchUI import WatchUI
+from WatchUI.WatchUI import WatchUI
```
Comment (radekBednarik) [inline WatchUI/__init__.py:1]: add newline

File: WatchUI/keywords/PDF.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+                and float(xy[3]) < float(y2)
+            ):
+                if xy_numbers == len(words_list):
+                    text += xy[4]
+                else:
+                    text += xy[4] + " "
+            xy_numbers += 1
+        return text
+
+    def should_exist_this_text(
```
Comment (radekBednarik) [inline WatchUI/keywords/PDF.py:90]: maybe "should_this_text_exist()"? makes more grammar sense ;)

File: WatchUI/keywords/Tesseract.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        Basics (object): interface class
+    """
+
+    def image_to_string(
+        self, path: str, oem: str="3", psm: str="3", language: str="eng", path_to_tesseract= ""
+    ) -> Union[bytes, str]:
+        """Converts provided image to string and returns.
+        Args:
+            path (str): path to the image
+            oem (str): [description]
```
Comment (radekBednarik) [inline WatchUI/keywords/Tesseract.py:21]: add description of the arguments please. I know, that I wrote this code, but I have no idea, what those arguments actually mean :)

File: WatchUI/keywords/Tesseract.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+        *coordinates,
+        oem: str = "3",
+        psm: str = "3",
+        language: str = "eng",
+        path_to_tesseract: str = "",
+    ) -> Union[list[str], str]:
+        """Convert to text only the selected area of the image and return.
+        Args:
+            path (str): path to the image
+            oem (str): [description]
```
Comment (radekBednarik) [inline WatchUI/keywords/Tesseract.py:51]: dtto

File: WatchUI/keywords/__init__.py [hunk 14]
```diff
@@ -0,0 +1,12 @@
+"""
+__init__ import methodst to WatchUI.py
```
Comment (radekBednarik) [inline WatchUI/keywords/__init__.py:2]: typo

File: examples/selenium/sample.robot [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+# This import need to be changed into the correct path (now using last version WUI from this repo)
+Library    ../../WatchUI/
+
+
+*** Test Cases ***
+Compare homapage with baseline
+    Open Browser                https://www.google.com    chrome
+    Capture Page Screenshot     baseline.png
+    Compare Image               baseline.png              baselines/homepage.png    ssim=0.95
+    Close Browser
```
Comment (radekBednarik) [inline examples/selenium/sample.robot:12]: add newline

**Review by radekBednarik — CHANGES_REQUESTED** [review]
review done. added comments, to individual files.
only one substantial question reqarding super(), otherwise small stuff.
did not checked the RF tests, I do not understand the RF logic that much, would be painful :)

File: .github/workflows/code-formatter.yml (hunk 1 above)
Comment (procesor2017) [reply .github/workflows/code-formatter.yml]: Done

File: .github/workflows/main.yml (hunk 2 above)
Comment (procesor2017) [reply .github/workflows/main.yml]: Done

File: .github/workflows/main.yml (hunk 3 above)
Comment (procesor2017) [reply .github/workflows/main.yml]: In the meantime, leave it as is, I will address the pipeline next Q.

File: .github/workflows/python-publish.yml (hunk 4 above)
Comment (procesor2017) [reply .github/workflows/python-publish.yml]: done

File: LICENSE (hunk 5 above)
Comment (procesor2017) [reply LICENSE]: Done

File: README.md (hunk 6 above)
Comment (procesor2017) [reply README.md]: Done

File: WatchUI/Ibasic/IBasic.py [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+                    html=True,
+                )
+            elif type_of_messages == "Error":
+                logger.error(
+                    "Images arent same</p><img src='" + path_to_image + "'/>", html=True
```
[the remark this answers is not in the document]
Comment (procesor2017) [reply WatchUI/Ibasic/IBasic.py]: Done

File: WatchUI/__init__.py (hunk 10 above)
Comment (janegermaier) [reply WatchUI/__init__.py]: Done

File: WatchUI/keywords/PDF.py (hunk 11 above)
Comment (janegermaier) [reply WatchUI/keywords/PDF.py]: I leave it, how we have now, because its historical keyword.

File: WatchUI/keywords/Tesseract.py (hunk 12 above)
Comment (janegermaier) [reply WatchUI/keywords/Tesseract.py]: Done

File: WatchUI/keywords/Tesseract.py (hunk 13 above)
Comment (janegermaier) [reply WatchUI/keywords/Tesseract.py]: Done

File: WatchUI/keywords/__init__.py (hunk 14 above)
Comment (janegermaier) [reply WatchUI/keywords/__init__.py]: Done

File: examples/selenium/sample.robot (hunk 15 above)
Comment (janegermaier) [reply examples/selenium/sample.robot:12]: Done

File: WatchUI/WatchUI.py (hunk 9 above)
Comment (janegermaier) [reply WatchUI/WatchUI.py]: Done

**Review by radekBednarik — APPROVED** [review]
LGTM

**Pull request merged**